libraries/javascript/eddystone-advertising/eddystone-advertising.js

```diff
@@ -0,0 +1,511 @@
+     @default
+   */
+  const EDDYSTONE_URL_FRAME_TYPE = 0x10;
+
+  /**
+     These schemes are used to encode/decode a URL Scheme prefix.
+     @see {@link https://github.com/google/eddystone/tree/master/eddystone-url#url-scheme-prefix|URL Scheme Prefix}
+     @private
+     @constant {string[]}
+     @default
+   */
+  const URL_SCHEMES = [
+    'http://www.',
+    'https://www.',
+    'http://',
+    'https://'
+  ];
+
+  /**
+     These codes are used to encode/decode a URL.
+     @see {@link https://github.com/google/eddystone/tree/master/eddystone-url#eddystone-url-http-url-encoding|HTTP URL Encoding}
+     @private
+     @constant {string[]}
+     @default
+   */
+  const URL_CODES = [
+    '.com/',
+    '.org/',
+    '.edu/',
+    '.net/',
+    '.info/',
+    '.biz/',
+    '.gov/',
+    '.com',
+    '.org',
+    '.edu',
+    '.net',
+    '.info',
+    '.biz',
+    '.gov'
+  ];
+
+  // The Eddystone-URL spec says to use 'us-ascii' encoding but not all
+  // browsers support it, so we use 'utf-8' which is a super set of
+  // 'us-ascii' and wildly supported. We later check the encoded characters
+  // are in the range allowed by the Eddystone-URL Spec.
+  // https://github.com/google/eddystone/tree/master/eddystone-url#eddystone-url-http-url-encoding
+  let encoder = new TextEncoder('utf-8');
+
+  /**
+     This class provides helper functions that relate to Eddystone-URL.
+     @see {@link https://github.com/google/eddystone/tree/master/eddystone-url|Eddystone-URL}
+     @class
+   */
+  class EddystoneURL {
+    /**
+       Constructs a valid Eddystone-URL service data from a URL and a Tx Power
+       value.
+       @see {@link https://github.com/google/eddystone/tree/master/eddystone-url#frame-specification|URL Frame Specification}
+       @param {string} url The URL to use in the service data.
+       @param {number} txPower The Tx Power to use in the service data.
+       @returns {number[]} The service data.
+       @throws {Error} If the Tx Power value is not in the allowed range. See
+       {@link https://github.com/google/eddystone/tree/master/eddystone-url#tx-power-level|Tx Power Level}.
+       @throws {Error} If the URL Scheme prefix is unsupported. For a list of
+       supported Scheme prefixes see
+       {@link https://github.com/google/eddystone/tree/master/eddystone-url#url-scheme-prefix|URL Scheme Prefix}
+       @throws {Error} If the URL contains an invalid character. For a list of
+       invalid characters see the Note in
+       {@link https://github.com/google/eddystone/tree/master/eddystone-url#eddystone-url-http-url-encoding|HTTP URL Encoding}
+     */
+    static constructServiceData(url, txPower) {
+      // Check that it's a valid Tx Power
+      if (txPower < -100 || txPower > 20) {
+        throw new Error('Invalid Tx Power value: ' + txPower);
+      }
+      let base_frame = [EDDYSTONE_URL_FRAME_TYPE, txPower];
+      Array.prototype.push.apply(base_frame, EddystoneURL.encodeURL(url));
```

Comment:
But that would make a whole new array. This way we just add to the already created array.